### ISSUE TITLE
Counter field validation still occurs after downgrading tsdb data stream

### ISSUE THREAD

**Issue by martijnvg** [elastic/elasticsearch#96170]
Tsdb is more picky about what aggregations are allowed to be executed on counter fields. For example a sum aggregation isn't allowed to run on a counter field. There is support to opt out of tsdb. However even after opting out of tsdb shard failures may still occur. This is because the data stream is downgraded but not any pre-existing tsdb indices.

Reproduction:

```
PUT _index_template/1
{
  "index_patterns": [
    "test*"
  ],
  "template": {
    "mappings": {
      "properties": {
          "my_field": {
              "time_series_dimension": true,
              "type": "keyword"
          },
          "another_field": {
            "type": "long",
            "time_series_metric": "counter"
          }
      }
    }
  },
  "data_stream": {}
}


POST test2/_doc?refresh
{
  "@timestamp": "2023-05-16T11:49:50.599Z",
  "my_field": "value",
  "another_field": 656
}

GET test2/_search
{
    "size": 0,
    "aggs": {
        "test": {
            "sum": {
                "field": "another_field"
            }
        }
    }
}


PUT _index_template/1
{
  "index_patterns": [
    "test*"
  ],
  "template": {
    "settings": {
      "index": {
        "mode": "time_series"
      }
    },
    "mappings": {
      "properties": {
          "my_field": {
              "time_series_dimension": true,
              "type": "keyword"
          },
          "another_field": {
            "type": "long",
            "time_series_metric": "counter"
          }
      }
    }
  },
  "data_stream": {}
}

POST test2/_rollover

POST test2/_doc?refresh
{
  "@timestamp": "2023-05-16T16:23:51.659Z",
  "my_field": "value",
  "another_field": 6566
}

GET test2/_search
{
    "size": 0,
    "aggs": {
        "test": {
            "sum": {
                "field": "another_field"
            }
        }
    }
}

PUT _index_template/1
{
  "index_patterns": [
    "test*"
  ],
  "template": {
    "settings": {
      "index": {
        "mode": null
      }
    },
    "mappings": {
      "properties": {
          "my_field": {
              "time_series_dimension": true,
              "type": "keyword"
          },
          "another_field": {
            "type": "long",
            "time_series_metric": "counter"
          }
      }
    }
  },
  "data_stream": {}
}

POST test2/_rollover

POST test2/_doc?refresh
{
  "@timestamp": "2023-05-16T16:23:51.659Z",
  "my_field": "value",
  "another_field": 6566
}

GET test2/_search
{
    "size": 0,
    "aggs": {
        "test": {
            "sum": {
                "field": "another_field"
            }
        }
    }
}
```

The expected behaviour is that the last search succeeds but it fails with a shard failure:

```
{
    "took": 4,
    "timed_out": false,
    "_shards": {
        "total": 3,
        "successful": 2,
        "skipped": 0,
        "failed": 1,
        "failures": [
            {
                "shard": 0,
                "index": ".ds-test2-2023.05.16-000002",
                "node": "e9y5C7zmRy6I-3j5aW1vEA",
                "reason": {
                    "type": "illegal_argument_exception",
                    "reason": "Field [another_field] of type [long] is not supported for aggregation [sum]"
                }
            }
        ]
    },
    "hits": {
        "total": {
            "value": 2,
            "relation": "eq"
        },
        "max_score": null,
        "hits": []
    },
    "aggregations": {
        "test": {
            "value": 7222
        }
    }
}
```

Note that this shard failure could be ignored.

**Comment by ruflin** [comment]
For all the "valid" aggregation, what happens with the tsdb indices after rollback? Assuming I look at 24h and after 12h I had a TSDB index for 1h and then rolled it back, will my aggs over the 24h show the correct results or will be there a missing entry for 1h? This is under the assumption, I use a max agg on a field that was a counter for 1h.

**Comment by martijnvg** [comment]
The `max` aggregation is supported on counter fields. And like any other supported aggregation for counter fields these will continue to work on tsdb data streams and downgraded data streams.

But for let's say a `sum` aggregation (and other unsupported aggs), which isn't supported aggregations on counter fields. These will not work on a tsdb data stream. And on a downgraded data stream, these aggregations will fail on backing indices that were created at the time the data stream was still in tsdb mode. This means no data/results will be returned for any searches have an unsupported aggregation in the time period a data stream was in tsdb mode. This will be returned as a shard error. Data/results will be returned from the time the data stream was downgraded to be a normal data stream.

**Comment by felixbarny** [comment]
> Data/results will be returned from the time the data stream was downgraded to be a normal data stream.

That sounds good to me. Just to double-check that I understood this correctly: Say you upgrade a data stream with existing metric data to TSDB, then downgrade after 1h and continue ingesting data. If you do a sum aggregation on a counter, the response will only miss data from that 1h period where the data stream was in TSDB mode. Is that correct?

**Comment by martijnvg** [comment]
> Say you upgrade a data stream with existing metric data to TSDB, then downgrade after 1h and continue ingesting data. If you do a sum aggregation on a counter, the response will only miss data from that 1h period where the data stream was in TSDB mode. Is that correct?

Yes. Additionally the search response will also report shard(s) failures, because sum aggregation was attempted on counter field in a tsdb backing index.

**Comment by martijnvg** [comment]
> Is there anything we need to change then?

I don't think so.

> We could ignore it but wouldn't that be even more surprising to users as there wouldn't be an explanation of why there's partially missing data?

Yes, we the shard level error should be returned.

**Comment by lalit-satapathy** [comment]
@felixbarny the end user case is explained the duplicate here issue [here](https://github.com/elastic/elasticsearch/issues/96169#issue-1712148089), need to reproduce and verify once.

**Comment by ruflin** [comment]
The key question for me is how Lens / Kibana behaves:

1. No error should be thrown
2. Ideally we would inform the user why there is a gap (enhancement)

**Comment by martijnvg** [comment]
> No error should be thrown

@dej611 Can you comment on what happens what ES returns shard failures in this particular case?

**Comment by martijnvg** [comment]
@dej611 I did a little bit of testing, to see what is happening now. If a data stream both contains tsdb and normal indices then the field caps api will not returns the `time_series_metric` attribute. Only if all backing indices are tsdb indices then the `time_series_metric` attribute is included in the field caps response. So I think that Lens would still allow all operations on counter fields on upgraded or downgraded data streams. But that means that when querying the data stream with something like a sum on counter field, then a shard failure will be returned by the Search API.

**Comment by ruflin** [comment]
> then a shard failure will be returned by the Search API.

What does Lens do with this shard failure from a user perspective?

**Comment by dej611** [comment]
In case of a dataView with mixed index types, then Lens won't classify the `counter` field as TSDB field because the `time_series_metric` won't be sent by the field caps API.
In this case Lens will render only part of the data, and report a warning message about the shard failure: all the data sent over will be used anyway to render the final visualization.

Here's an example of the Lens rendering in this situation (I've used the repro steps above, then added a regular data stream `test3`, both in the same dataView):
<img width="944" alt="Screenshot 2023-05-23 at 15 06 41" src="https://github.com/elastic/elasticsearch/assets/924948/e7075450-ad2f-4309-86ee-e0775ce6ae2b">
<img width="949" alt="Screenshot 2023-05-23 at 15 06 49" src="https://github.com/elastic/elasticsearch/assets/924948/01bfa536-7f6d-4c7c-88e9-eee5ca0376a6">
<img width="756" alt="Screenshot 2023-05-23 at 15 08 19" src="https://github.com/elastic/elasticsearch/assets/924948/df936a84-f4a5-4825-a5e3-57ce335362dc">

**Comment by martijnvg** [comment]
This is the view I get in Kibana based on the above reproduction:

<img width="1129" alt="image" src="https://github.com/elastic/elasticsearch/assets/580421/793b7f92-127c-4193-b3ec-d90892b4e937">

**Comment by dej611** [comment]
> Does that imply that any aggregations that exclusively work with TSDB can only be used if the data stream has always been in TSDB mode or if the non-TSDB backing indices are deleted/aged out?

If `time_series_metric` is not sent over to Lens, then a `counter` field is just identified as regular `number` field type and all operations can be performed as above. There are no TSDB specific operations for a `counter` field yet, Lens only disables some operations for it, won't enable new ones.
When the `rate` TSDB agg will be integrated in Lens, that would be completely transparent for the user, which will still use the `Counter Rate` operation for his prospective.

**Comment by lalit-satapathy** [comment]
@agithomas has the simple test done, where the behaviour is summarised for **Non-TSDB to TSDB to Non-TSDB** in a visualisation with non-supported aggregation on counters. This captures the essence of the custom dashboard scenario described in the [issue](https://github.com/elastic/elasticsearch/issues/96169#issue-1712148089). @agithomas please add the behaviour here.

**Comment by agithomas** [comment]
**Please find below, the changes in visualisation when the transition of Non-TSDB -> TSDB -> Non-TSDB occur.**

<img width="1488" alt="image" src="https://github.com/elastic/elasticsearch/assets/101976829/abc170b2-862f-4656-8ad6-268fb3393648">

**Comment by martijnvg** [comment]
This issue was discussed in yesterday's tsdb integration sync. The fact that visualisation that are not allowed on counter field show gaps for the time the data stream was in tsdb mode is ok. However instead of displaying that one or more shards failed we like to display this fact differently. First step is that on the ES side we can include in the error message that this is about a counter field. Currently the field type is included in the error message (but not that it is a counter metric field). Then on the Kibana side logic can be added that displays this kind of situations differently (less sever than with a shard failure).

**Comment by martijnvg** [comment]
#93545 will address the error on the Elasticsearch side.

**Comment by martijnvg** [comment]
This is addressed via #97974
